Refactor to use singledispatch

BenjaminBossan opened the pull request:
This is a proposal to refactor the code to use `singledispatch`.

## Description

So far, we have written a custom dispatch logic by walking the mro until we find a suitable object. Now we use a `singledispatch` decorator instead which under the hood does something very similar but is based on the builtin `singledispatch` decorator.

### Pros

It is no longer necessary to litter the code with `get_state_method` and `get_instance_method` calls, as that is resolved through the decorator. Also, we can assume that the `singledispatch` decorator is battle tested and thus more robust than our custom implementation. Moreover, it seems to be using some form of caching, so maybe it's also faster. Finally, since we no longer need to resolve the methods, the local imports are gone.

### Cons

Unfortunately, we could not rely completely on the built in `singledispatch` decorator. The reason is the following:

Usually, we want to dispatch on the class of the object. However, when we call get_instance, the object is *always* a dict, which invalidates the dispatch. Therefore, we change the dispatcher to dispatch on the instance, not the class. By default, we just use the class of the instance being passed, i.e. we do exactly the same as in the original implementation. However, if we encounter a state dict, we resolve the actual class from the state dict first and then dispatch on that class.

Another con is that we have to decorate all the functions that we plan on using. Probably, this could be solved similarly to the previous implementation by dynamically importing the functions and decorating them on the fly, I have not tried.

BenjaminBossan committed:
Refactor to use singledispatch

This is a proposal to refactor the code to use singledispatch.

Description:

So far, we have written a custom dispatch logic by walking the mro until
we find a suitable object. Now we use a singledispatch decorator instead
which under the hood does something very similar but is based on the
builtin singledispatch decorator.

Pros:

It is no longer necessary to litter the code with get_state_method and
get_instance_method calls, as that is resolved through the decorator.
Also, we can assume that the singledispatch decorator is battle tested
and thus more robust than our custom implementation. Finally, it seems
to be using some form of caching, so maybe it's also faster. Finally,
since we no longer need to resolve the methods, the local imports are
gone.

Cons:

Unfortunately, we could not rely completely on the built in
singledispatch decorator. The reason is the following:

Usually, we want to dispatch on the class of the object. However, when
we call get_instance, the object is *always* a dict, which invalidates
the dispatch. Therefore, we change the dispatcher to dispatch on the
instance, not the class. By default, we just use the class of the
instance being passed, i.e. we do exactly the same as in the original
implementation. However, if we encounter a state dict, we resolve the
actual class from the state dict first and then dispatch on that class.

Another con is that we have to decorate all the functions that we plan
on using. Probably, this could be solved similarly to the previous
implementation by dynamically importing the functions and decorating
them on the fly, I have not tried.

BenjaminBossan commented:
Please review @adrinjalali

adrinjalali commented on skops/io/_general.py:
```diff
@@ -115,6 +111,8 @@ def function_get_state(obj, dst):
     return res
 
 
+@get_instance.register(np.ufunc)
```
The issue with this solution is that we're moving the _register_ action to where the method is defined. I'm not sure how to make this work with multiple versions for instance, and it's also now harder to see all methods related to a module (which I guess is not a big deal).

BenjaminBossan replied:
Yes, for sure. I think it can be avoided however, since decorators are just function calls + assignments in the end. We could probably dynamically register functions in a similar fashion as in the original implementation by having the modules present a dict/list of types and their corresponding methods and then have something like:

```python
for cls, method in skops._numpy.get_state_methods.items():
    get_state.register(cls)(method)
```

I haven't tested that yet.

adrinjalali replied:
I think that'd be really cool, but we don't have to have the whole `get_state_methods` thing there, we could just explicitly register the things which are not in the same module's file.

BenjaminBossan replied:
I'll give it a try once this is merged, don't want to complicate things too much.

adrinjalali commented on skops/io/_utils.py:
```diff
@@ -1,4 +1,135 @@
 import importlib
+from functools import (  # type: ignore
+    GenericAlias,
+    _find_impl,
```
I'm somehow more comfortable using private sklearn methods rather than private python methods lol. Let's see how much headache this would give us.

BenjaminBossan replied:
I didn't check the implementations in detail, but in the worst case, we could probably copy the code of `_find_impl` here. I wouldn't do it now since it depends on another private function which in turn depends on another private function :D

adrinjalali commented on skops/io/_general.py:
```diff
@@ -115,6 +111,8 @@ def function_get_state(obj, dst):
+@get_instance.register(FunctionType)
 def function_get_instance(obj, src):
     loaded = _import_obj(obj["__module__"], obj["__content__"])
     return loaded
```
remove lines bellow here?

BenjaminBossan replied:
good catch

adrinjalali reviewed (commented):
This is simply BEAUTIFUL. Other than my slight concerns, we should test this with different python versions. Could you please remove 

```
    if: "github.repository == 'skops-dev/skops'"
```

from `build-test.yml` to test CI here? (I think it'd work, not sure)

Otherwise this is brilliant!

BenjaminBossan committed:
Reviewer comments:

- Allow CI to run
- Remove unused code

BenjaminBossan committed:
Fix for failing GenericAlias import

Was introduced in Python 3.9, added fix for earlier versions.

BenjaminBossan commented:
@adrinjalali I addressed your points, please review again.

One CI run has timed out but for some reason I cannot restart it, maybe because the PR is against your fork?

adrinjalali commented on .github/workflows/build-test.yml:
```diff
@@ -12,7 +12,6 @@ jobs:
   pytest:
 
     runs-on: ${{ matrix.os }}
-    if: "github.repository == 'skops-dev/skops'"
```
happy to merge, but we need to put this one back first.

adrinjalali reviewed (approved):
Thanks @BenjaminBossan

adrinjalali commented:
I added the line in the main PR.
